### PR TITLE
Feat: visual component for Project certificates

### PR DESCRIPTION
**Pull request by rkpattnaik780**
### Type of change

- Enhancement / new feature
- Bugfix

### Description

The visual components required for Project Certificate Lists have been created. It addresses the following: 

* Certificates can be displayed in cards.
* A form to create/edit certificates.
* Auto-population of form fields upon edit.
* Toggling between view and edit mode.

### Screenshots

Screenshot from 2020-05-27 12-21-57

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [ ] Write tests and make sure they pass
- [x] Resolve all warnings
- [x] Update license
- [ ] Write test cases for null checks in Unit Tests
- [ ] Reference relevant issue(s) and close them after merging

### CODE REVIEW

**Comment by rkpattnaik780** [comment]
We should lay out a guideline for naming ids when we are rendering multiple elements with map.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            <Title
+              headingLevel="h1"
+              size="md"
+              id={`cc-auto-provision-title-${id}`}
+            >
+              <b>{getCertificateLabelByKey("auto-provisioning-enabled")}</b>
+            </Title>
+          </GridItem>
+          <GridItem span={10} className={rowMargin}>
+            <Switch
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx:111]: Use custom SwitchWithToggle component

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+              id={`cc-auto-provision-switch-${id}`}
+              label="Enabled"
+              labelOff="Disabled"
+              readOnly
+              isChecked={certificate["auto-provisioning-enabled"]}
+            />
+          </GridItem>
+          <GridItem span={2}>
+            <Title headingLevel="h1" size="md" id={`cc-algorithm-title-${id}`}>
+              <b>{getCertificateLabelByKey("algorithm")}</b>
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx:121]: We have aleady one fuction for same purpose in top level utils i.e. getLabelByKey. You can use that and add missing key level as you need.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+  setOnEditMode: React.Dispatch<React.SetStateAction<boolean>>;
+  id: string;
+}
+
+export const CertificateForm: React.FunctionComponent<ICertificateFormProps> = ({
+  setOnEditMode,
+  certificate,
+  id
+}) => {
+  const [subjectDn, setSubjectDn] = useState<string>();
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx:44]: can you try to store all the values in single state as a object.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  const [publicKey, setPublicKey] = useState<string>();
+  const [algorithm, setAlgorithm] = useState<string>();
+  const [notBefore, setNotBefore] = useState<string>();
+  const [notAfter, setNotAfter] = useState<string>();
+  const [autoProvision, setAutoProvision] = useState<boolean>();
+
+  useEffect(() => {
+    // code to initialize the form if it is being used for edit
+    if (certificate) {
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx:53]: Not sure, why do you need useEffect and if things are depends on certificate, you must pass certificate as dependancy.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+      setSubjectDn(subject_dn);
+      setPublicKey(public_key);
+      setAlgorithm(algo);
+      setNotBefore(not_before);
+      setNotAfter(not_after);
+      setAutoProvision(auto_provision);
+    }
+  }, []);
+
+  const onSubjectDnChange = (value: string) => {
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx:72]: All below function doing same thing. We don't need so many function for same purpose.
You can create single fucntion and get value and find elementName and update state for specific field.

For reference
const [formData,setFormData]=useState({filed1:"",field2:""});

   const onChangeInput=(value:string,evt:any)=>{
      const elementName=evt?.target?.name;
      const newFormData={...formData};
      (newFormData as any)[elementName]=value;
      setFormData(newFormData);      
   }

use  onChangeInput function which will also solve multiple state issue.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/ProjectCertificateToolbar.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import {
+  DataToolbar,
+  DataToolbarContent,
+  DataToolbarItem,
+  Button,
+  Switch
+} from "@patternfly/react-core";
+
+export interface IProjectCertificateToolbarProps {
+  setShowCertificateForm: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/ProjectCertificateToolbar.tsx:16]: is there any specific reason for using type React.Dispatch<React.SetStateAction<boolean>> ?

File: console/console-init/ui/src/modules/project-detail/utils/constant.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright 2020, EnMasse authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+import { ISelectOption } from "utils";
+
+export const algorithmTypeOptions: ISelectOption[] = [
+  { key: "RSA", label: "RSA", value: "RSA" },
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/utils/constant.ts:9]: try to keep keys in small case i.e. key:"rsa", key:"ec"

File: console/console-init/ui/src/modules/project-detail/utils/utils.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright 2020, EnMasse authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+export const getCertificateLabelByKey = (key: string) => {
```
Comment (ajaypratap003) [inline console/console-init/ui/src/modules/project-detail/utils/utils.ts:6]: we don't need separate label function , we have aleady function in top lebel utils , getLabelBykey. you can add missing key label in that functuon.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/ProjectCertificateToolbar.tsx (hunk 6 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/ProjectCertificateToolbar.tsx:16]: I am passing the setter function returned by `useSate` as a prop.

File: console/console-init/ui/src/modules/project-detail/utils/utils.ts (hunk 8 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/utils/utils.ts:6]: Created a separate function as having a common label function at top level might make it overpopulated. Should I use the common one anyway?

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx (hunk 4 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx:53]: There are two use cases of this form - one is to edit an existing certificate and another to create a new certificate. The second case doesn't need a prop.

File: console/console-init/ui/src/modules/project-detail/utils/constant.ts (hunk 7 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/utils/constant.ts]: done

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx (hunk 5 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx]: Done. We should have it in other forms as well.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx (hunk 3 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx]: Done.

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx (hunk 2 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx]: done

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx (hunk 1 above)
Comment (rkpattnaik780) [reply console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateCard.tsx:111]: Done.

File: console/console-init/ui/src/utils/utils.ts [hunk 9]
```diff
@@ -280,7 +280,11 @@ const getLabelByKey = (key: string) => {
     "pwd-hash": "Password",
     hashed_password: "Password",
     psk: "PSK",
-    "x-509": "X-509 certificate"
+    "x-509": "X-509 certificate",
+    "subject-dn": "Subject-dn",
+    "public-key": "Public key",
+    "auto-provisioning-enabled": "Auto-provision",
+    algorithm: "Algorithm"
```
Comment (ajaypratap003) [inline console/console-init/ui/src/utils/utils.ts:287]: I think we don't need this kind of key label (algorithm: "Algorithm") for first letter cap. You can handle by using css textTransform: "capitalize"

File: console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+  useEffect(() => {
+    // code to initialize the form if it is being used for edit
+    certificate && setCertificateFormData(certificate);
+  }, []);
+
+  const onCertificateFormDataChange = (
+    value: string | boolean,
+    event: React.FormEvent<HTMLInputElement>
+  ) => {
+    const cerftificateField = event?.target?.name;
```
Comment (anukritijha) [inline console/console-init/ui/src/modules/project-detail/components/ProjectCertificateList/CertificateForm.tsx:72]: typo *cerftificateField

**Pull request merged**